File: web/src/services/dashboards.ts
// Copyright 2023 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import http from "./http";

const dashboards = {
  list: (
    page_num: number,
    page_size: number,
    sort_by: string,
    desc: boolean,
    name: string,
    organization: string,
    folderId: string,
    title: string
  ) => {
    const params: any = {
      page_num,
      page_size,
      sort_by,
      desc,
      name,
    };
    if (folderId) {
      params.folder = folderId;
    }
    // Only add title if it is provided
    if (title) {
      params.title = title;
    }
    return http().get(`/api/${organization}/dashboards`, { params });
  },
  create: (organization: string, data: any, folderId?: string) => {
    return http().post(
      `/api/${organization}/dashboards?folder=${folderId ?? "default"}`,
      data,
      { headers: { "Content-Type": "application/json; charset=UTF-8" } }
    );
  },
  delete: (organization: string, dashboardID: string, folderId?: string) => {
    return http().delete(
      `/api/${organization}/dashboards/${dashboardID}?folder=${folderId ?? "default"
      }`
    );
  },
  bulkDelete: (organization: string, data: any, folder?: string) => {
    let url = `/api/${organization}/dashboards/bulk`;
    if (folder) {
      url += `?folder=${folder}`;
    }
    return http().delete(url, { data });
  },
  get_Dashboard: (org_identifier: string, dashboardID: string, folderId?: string) => {
    return http().get(`/api/${org_identifier}/dashboards/${dashboardID}?folder=${folderId ?? "default"}`);
  },
  save: (
    organization: string,
    dashboardID: string,
    data: any,
    folderId?: string,
    hash?: any
  ) => {
    return http().put(
      `/api/${organization}/dashboards/${dashboardID}?folder=${folderId ?? "default"
      }&hash=${hash}`,
      data,
      { headers: { "Content-Type": "application/json; charset=UTF-8" } }
    );
  },
  list_Folders: (organization: string) => {
    return http().get(`/api/v2/${organization}/folders/dashboards`);
  },
  new_Folder: (organization: string, data: any) => {
    return http().post(`/api/v2/${organization}/folders/dashboards`, data, {
      headers: { "Content-Type": "application/json; charset=UTF-8" },
    });
  },
  edit_Folder: (organization: string, folderId: any, data: any) => {
    return http().put(`/api/v2/${organization}/folders/dashboards/${folderId}`, data, {
      headers: { "Content-Type": "application/json; charset=UTF-8" },
    });
  },
  delete_Folder: (organization: string, folderId: any) => {
    return http().delete(`/api/v2/${organization}/folders/dashboards/${folderId}`);
  },
  move_Dashboard: (organization: string, dashboardIds: string[], from: string, dstFolderId: string) => {
    return http().patch(
      `/api/${organization}/dashboards/move?folder=${from}`,
      {
        "dashboard_ids": dashboardIds,
        "dst_folder_id": dstFolderId
      },
      { headers: { "Content-Type": "application/json; charset=UTF-8" } }
    );
  },
};

export default dashboards;

File: src/common/migration/dashboards.rs
// Copyright 2025 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use config::{
    TIMESTAMP_COL_NAME,
    meta::{
        dashboards::Dashboard,
        stream::{DistinctField, StreamSettings, StreamType},
    },
    utils::time::now_micros,
};
use hashbrown::HashMap;
use infra::{
    db as infra_db,
    schema::get_settings,
    table::{
        dashboards,
        distinct_values::{self, DistinctFieldRecord, OriginType},
    },
};

use crate::service::{dashboards::get_query_variables, stream::save_stream_settings};

type SettingsCache = HashMap<(String, String, StreamType), StreamSettings>;

/// This takes query variables from the dashboard and inserts them
/// in the distinct_values table as well as the stream settings
async fn add_distinct_from_dashboard(
    org_id: &str,
    dashboard: &Dashboard,
    settings_cache: &mut SettingsCache,
) -> Result<(), anyhow::Error> {
    let dashboard_id = dashboard
        .dashboard_id()
        .ok_or_else(|| anyhow::anyhow!("Dashboard ID is missing"))?;

    let variables = get_query_variables(Some(dashboard));

    for ((stream, stype), fields) in variables {
        let cache_key = (org_id.to_string(), stream.clone(), stype);

        let stream_settings = match settings_cache.get_mut(&cache_key) {
            Some(s) => s,
            None => {
                let settings = get_settings(org_id, &stream, stype)
                    .await
                    .unwrap_or_default();
                settings_cache.insert(cache_key.clone(), settings);
                settings_cache.get_mut(&cache_key).unwrap()
            }
        };

        for f in fields {
            if f == "count" || f == TIMESTAMP_COL_NAME {
                // these two are reserved for oo use, so cannot be added as
                // separate distinct fields
                continue;
            }
            let record = DistinctFieldRecord::new(
                OriginType::Dashboard,
                dashboard_id,
                org_id,
                &stream,
                stype.to_string(),
                &f,
            );

            let temp = DistinctField {
                name: f,
                added_ts: now_micros(),
            };

            if !stream_settings.distinct_value_fields.contains(&temp) {
                stream_settings.distinct_value_fields.push(temp);
            }

            distinct_values::add(record).await?;
        }
    }
    Ok(())
}

// due to the order we init the resources, this migration will always run after
// the sea-orm migrations, so we can be certain that dashboard folders are migrated here.
pub async fn run() -> Result<(), anyhow::Error> {
    // load dashboards list
    let db = infra_db::get_db().await;

    // The following for-loop, which assigns all dashboards in the meta table to
    // a default folder, is deprecated as of v0.14.0 and will be removed in
    // v0.17.0.
    let db_key = "/dashboard/".to_string();
    let data = db.list(&db_key).await?;
    for (key, val) in data {
        let local_key = key.strip_prefix('/').unwrap_or(&key);
        let len = local_key.split('/').collect::<Vec<&str>>().len();
        if len > 3 {
            continue;
        }
        let new_key = key.replace("/dashboard/", "/dashboard/default/");
        match db.put(&new_key, val, infra_db::NO_NEED_WATCH, None).await {
            Ok(_) => {
                let _ = db.delete(&key, false, infra_db::NO_NEED_WATCH, None).await;
                println!("Migrated dashboard: {} successfully", key);
            }
            Err(_) => {
                println!("Failed to migrate dashboard: {}", new_key);
            }
        }
    }

    // The logic that follows is dependent on the dashboards table created in
    // the SeaORM migration list. Unlike the logic above, it will not be
    // deprecated in v0.14.0, though we could probably move it into its own
    // migration script in the SeaORM migration list so that it is tracked by
    // SeaORM.
    if distinct_values::len().await? > 0 {
        log::info!("dashboard distinct values migration already done.");
    } else {
        let dashboards = dashboards::list_all().await?;
        let mut settings_cache: SettingsCache = HashMap::new();

        for (org, dashboard) in dashboards {
            match add_distinct_from_dashboard(&org, &dashboard, &mut settings_cache).await {
                Ok(_) => log::info!(
                    "dashboard {} migrated for distinct values",
                    dashboard.dashboard_id().unwrap()
                ),
                Err(e) => log::error!(
                    "failed to process variables from dashboard {} for distinct values : {e}",
                    dashboard.dashboard_id().unwrap()
                ),
            }
        }

        for ((org_id, stream, stype), settings) in settings_cache {
            match save_stream_settings(&org_id, &stream, stype, settings).await {
                Ok(_) => {}
                Err(e) => {
                    log::error!(
                        "failed to save updated stream settings when migrating for distinct values {org_id}/{stream} {stype} :{e} "
                    );
                }
            }
        }

        log::info!("dashboard distinct value migration completed");
    }

    Ok(())
}

macro_rules! _get_variables {
    ($map:ident, $dash:ident) => {
        if let Some(vars) = &$dash.variables {
            for v in vars.list.iter() {
                if let Some(ref qd) = v.query_data {
                    $map.entry((qd.stream.clone(), qd.stream_type))
                        .or_default()
                        .push(qd.field.clone());
                }
            }
        }
    };
}
